lightning/src/ln/outbound_payment.rs

```diff
@@ -282,6 +282,45 @@ impl PendingOutboundPayment {
 		insert_res
 	}
 
+	pub(super) fn insert_from_monitor_on_startup(
+		&mut self, session_priv: [u8; 32], path: &Path, payment_hash: PaymentHash,
+		best_block_height: u32
+	) -> bool {
+		match self {
+			PendingOutboundPayment::AwaitingInvoice { .. } |
+				PendingOutboundPayment::InvoiceReceived { .. } => {
+					// If we've reached this point, it means we initiated a payment to a BOLT 12 invoice and
+					// locked the htlc(s) into the `ChannelMonitor`(s), but failed to persist the
+					// `ChannelManager` after transitioning from this state to `Retryable` prior to shutdown.
+					// Therefore, we need to move this payment to `Retryable` now to avoid double-paying if
+					// the recipient sends a duplicate invoice.
+					//
+					// We leave many fields as `None` if they are only used for retries, since we'll never
+					// retry on startup. See this method's callsite in `ChannelManager::read`, but this
+					// mirrors the behavior if an HTLC in a monitor is entirely missing from
+					// `OutboundPayments`.
+					*self = Self::Retryable {
```

Comment:
I think this can be DRY'd with what we have in `ChannelManager::read`. Working on some updates.